### ISSUE TITLE
hostvars variables are not recursively templated in template module

### ISSUE THREAD

**Issue by mbialozor** [ansible/ansible#24979]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

- Template/Vars

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, May 10 2017, 15:12:38) [GCC 7.1.1 20170503 (Red Hat 7.1.1-1)]
```

##### CONFIGURATION

```
[defaults]
ask_sudo_pass = True
hostfile = hosts
hash_behaviour = merge

[privilege_escalation]
become_method = sudo
```

##### OS / ENVIRONMENT

```
Fedora release 26 (Twenty Six)
Linux nostromo.local 4.11.2-300.fc26.x86_64 #1 SMP Mon May 22 16:15:55 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

##### SUMMARY

I'm creating balancer configuration for application cluster. This configuration file should have entries for application instances master host and all slaves. Application instances are named using base name (for example: app1_inst) and number incremented on each cluster host (for example: app1_inst1 and app2_inst1 on master, app1_inst2 and app2_inst2 on first slave, etc.). I'm putting all slaves to dedicated group, then using host index from inside this group (incremented by 1 as group index starts from 0) to create application instance name. The case is that I want to just use _{{ inst.name }}_ in template for both situations, for master hosts and for slaves. That's why I use _{%- for inst in hostvars[slave]['instances'] %}_ loop. It looks like variable _{{ ansible_play_batch.index(inventory_hostname) + 1 }}_ is not properly expanded for slaves.

I use _{{ ansible_play_batch.index(inventory_hostname) + 1 }}_ construction in several places on both master and slaves, without any issues. It looks like this construction works well when it refers to current host. There may be some problem when using it with conjunction with _hostvars_.

According to official documentation (http://docs.ansible.com/ansible/playbooks_variables.html):

> - play_hosts has been deprecated in 2.2, it was the same as the new ansible_play_batch variable.
>
> - ansible_play_hosts is the full list of all hosts still active in the current play.
>
> - ansible_play_batch is available as a list of hostnames that are in scope for the current ‘batch’ of the play. The batch size is defined by serial, when not set it is equivalent to the whole play (making it the same as ansible_play_hosts).

I observe the same issue using every of those three variables:
- play_hosts
- ansible_play_hosts
- ansible_play_batch

##### STEPS TO REPRODUCE

**hosts**
```yaml
[cluster]
cluster-ma.domain.local cluster_role=master
cluster-sl01.domain.local cluster_role=slave
```
**group_vars/cluster**
```yaml
instances:
  - name: 'app1_inst{{ ansible_play_batch.index(inventory_hostname) + 1 }}'
    basename: app1_inst
    group: app1-group
    autostart: yes    
  - name: 'app2_inst{{ ansible_play_batch.index(inventory_hostname) + 1 }}'
    basename: app2_inst
    group: app2-group
    autostart: no
```
**pre_tasks** (group hosts by role)
```yaml
- name: Group all hosts by role in cluster
  group_by:
    key: 'cluster_{{ cluster_role }}s'
```
**httpd.conf.j2** (create config file from template, only on master, which includes entries for master and all slaves):
```yaml
# Entry for master host instances on master host
{% for inst in instances %}
BalancerMember {{ inventory_hostname }} {{ inst.name }}
{% endfor %}

# Entries for all instances on all slave hosts
{% if cluster_role == "master" %}
{% for slave in groups['cluster_slaves'] %}
{% for inst in hostvars[slave]['instances'] %}
BalancerMember {{ slave }} {{ inst.name }}
{% endfor %}
{% endfor %}
{% endif %}
```

##### EXPECTED RESULTS

```
BalancerMember cluster-ma.domain.local route=app1_inst1
BalancerMember cluster-sl01.domain.local route=app1_inst2

BalancerMember cluster-ma.domain.local route=app2_inst1
BalancerMember cluster-sl01.domain.local route=app2_inst2
```

##### ACTUAL RESULTS

```
BalancerMember cluster-ma.domain.local route=app1_inst1
BalancerMember cluster-sl01.domain.local route=app1_inst{{ ansible_play_batch.index(inventory_hostname) + 1 }}

BalancerMember cluster-ma.domain.local route=app2_inst1
BalancerMember cluster-sl01.domain.local route=app2_inst{{ ansible_play_batch.index(inventory_hostname) + 1 }}
```

##### WORKING HACKS

1. Using in group variables:
```yaml
- name: 'app1_inst{{ groups.cluster.index(inventory_hostname) + 1 }}'
```
instead of:
```yaml
- name: 'app1_inst{{ ansible_play_batch.index(inventory_hostname) + 1 }}'
```
seems to be working as intended.

2. Using _{{ groups.cluster_slaves.index(slave) + 2 }}_ inside template:
```yaml
# Entry for master host instances on master host
{% for nodfe==e in instances %}
BalancerMember {{ inventory_hostname }} {{ node.name }}
{% endfor %}

# Entries for all instances on all slave hosts
{% if cluster_role == "master" %}
{% for slave in groups['cluster_slaves'] %}
{% for node in hostvars[slave]['instances'] %}
BalancerMember {{ slave }} route={{ node.basename }}{{ groups.cluster_slaves.index(slave) + 2 }}
{% endfor %}
{% endfor %}
{% endif %}
```
also gives me expected results.

**Comment by jctanner** [comment]
@mbialozor There's a lot of text to parse in this description, but If I understand correctly, the crux of the problem is that you want to use magic vars from the play to construct group vars?

For example:
```
name: 'app2_inst{{ ansible_play_batch.index(inventory_hostname) + 1 }}'
```

You want that to template properly inside a group vars file and not actually get templated until the play executes?

needs_info

**Comment by mbialozor** [comment]
@jctanner I use magic vars inside group vars. I need "index" number for constructing instance name on every cluster node (for example app1_inst1 on first node, app1_inst2 on second node, etc.), which index is taken from host index inside play. I need it to be templated correctly inside template file (something.j2), not directly in group vars.

So, inside group vars I use construction:
```
name: 'app2_inst{{ ansible_play_batch.index(inventory_hostname) + 1 }}'
```
Then in my template I use:
```
BalancerMember {{ inventory_hostname }} {{ inst.name }}
```
I expect something like this:
```
BalancerMember cluster-sl01.domain.local route=app1_inst2
```
But I get:
```
BalancerMember cluster-sl01.domain.local route=app1_inst{{ ansible_play_batch.index(inventory_hostname) + 1 }}
```
I looks like magic var is not treated here like a variable but just like a plain text.

As I said, host index taken from group (not play) is treated like a variable and gives me expected result.
```
name: 'app1_inst{{ groups.cluster.index(inventory_hostname) + 1 }}'
```

**Comment by mbialozor** [comment]
@jctanner using:
```
{{ ansible_play_batch.index(inventory_hostname) }}
```
is a little bit more universal than:
```
{{ groups.cluster.index(inventory_hostname) }}
```
as second construction requires giving group name.

**Comment by tchernomax** [comment]
I think I experience the same issue and manage to narrow it a bit.

Here is a minimal playbook/configuration to show and reproduce the problem.
```
max@laptop % ls
ansible.cfg  hosts  test.yml
max@laptop % tail -n+1 *
==> ansible.cfg <==
[defaults]
inventory           = hosts

==> hosts <==
localhost

[all:vars]
ansible_connection=local
ansible_python_interpreter=/usr/bin/python2

a=toto
b="{{a}}"
c="{{d}}"

==> test.yml <==
---
- name: all
  hosts: localhost
  gather_facts: false

  vars:
    d: titi

  tasks:

    - debug:
        var: b

    - debug:
        var: hostvars['localhost']['b']

    - debug:
        var: c

    - debug:
        var: hostvars['localhost']['c']

max@laptop % ansible-playbook test.yml

PLAY [all] ***********************************************************************************************

TASK [debug] *********************************************************************************************
ok: [localhost] => {
    "b": "toto"
}

TASK [debug] *********************************************************************************************
ok: [localhost] => {
    "hostvars['localhost']['b']": "toto"
}

TASK [debug] *********************************************************************************************
ok: [localhost] => {
    "c": "titi"
}

TASK [debug] *********************************************************************************************
ok: [localhost] => {
    "hostvars['localhost']['c']": "{{d}}"
}

PLAY RECAP ***********************************************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0
```
I keep the "a" and "b" variable to show the bug doesn't appear when variables are defined in the same file.

I am using ansible 2.3.2.0:
```
ansible 2.3.2.0
  config file = /home/max/Telechargements/ansible2/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jul 21 2017, 03:24:34) [GCC 7.1.1 20170630]
```

PS: the result is exactly the same with "gather_facts: true" (in case you wonder)

**Comment by westsouthnight** [comment]
Im so sorry but after read that topic i thinks its looks like closed two days ago bug from me.

@mbialozor please look to https://github.com/ansible/ansible/issues/16004

and if its looks like that, write about it please. I see bad behavior in last year and have difficult to split my dicts correctly if for different groups of hosts we have in one inventory dicts with same names. 

Always gets last!

**Comment by tchernomax** [comment]
@westsouthnight for my part the issue is still present in the devel branch.
```
(venv) max@test % ansible --version
ansible 2.5.0
  config file = /home/max/Telechargements/ansible.cfg
  configured module search path = ['/home/max/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/max/Telechargements/aaaa/lib/python3.6/site-packages/ansible
  executable location = /home/max/Telechargements/aaaa/bin/ansible
  python version = 3.6.3 (default, Oct 24 2017, 14:48:20) [GCC 7.2.0]

(venv) max@test % ansible-playbook test.yml

PLAY [all] **************************************************************************************

TASK [debug] **********************************************************************************
ok: [localhost] => {
    "b": "toto"
}

TASK [debug] **********************************************************************************
ok: [localhost] => {
    "hostvars['localhost']['b']": "toto"
}

TASK [debug] **********************************************************************************
ok: [localhost] => {}

TASK [debug] **********************************************************************************
ok: [localhost] => {
    "hostvars['localhost']['c']": "{{d}}"
}

PLAY RECAP **********************************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0
```

But I am not sure @mbialozor and I experienced exactly the same issue (If that's not the case I will create my own github issue).
@mbialozor can you check with your conf ?

**Comment by mbialozor** [comment]
This seems to be still an issue on Ansible 2.5.2.

**Comment by westsouthnight** [comment]
4 years guys! incredible @bcoca !!!

**Comment by bcoca** [comment]
possible fix #72419, closing as dupe as per above
